### ISSUE TITLE
Rails instrumentation breaks on Rails v4.2.11

### ISSUE THREAD

**Issue by owais** [open-telemetry/opentelemetry-ruby#569]
-----------------
**Rails: 4.2.11**
**Ruby: 2.6.6**
**opentelemetry-instrumentation-rails: 0.12.0**

Example project where issue can be reproduced: https://github.com/owais/rails-otel

--------------

I tried OpenTelemetry Ruby on a Rails project and ran into an issue with the rails instrumentation gem. Looks like the instrumentation patches ActionController Metal's dispatch method and expects three arguments, `name`, `request` and `response` but only `name` and `request` is passed into it. This results in the below exception. If I patch the instrumentation to make the third argument (response) optional, everything works as expected.

--------------------

<img width="1063" alt="image" src="https://user-images.githubusercontent.com/46186/105426393-10ed9e80-5c71-11eb-9af1-780da3c5a908.png">

<details><summary>Full stack trace</summary>

```
vendor/bundle/gems/opentelemetry-instrumentation-rails-0.12.0/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb:14:in `dispatch'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_controller/metal/rack_delegation.rb:13:in `dispatch'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_controller/metal.rb:237:in `block in action'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/routing/route_set.rb:74:in `dispatch'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/routing/route_set.rb:43:in `serve'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/journey/router.rb:43:in `block in serve'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/journey/router.rb:30:in `each'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/journey/router.rb:30:in `serve'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/routing/route_set.rb:817:in `call'
vendor/bundle/gems/rack-1.6.13/lib/rack/etag.rb:24:in `call'
vendor/bundle/gems/rack-1.6.13/lib/rack/conditionalget.rb:25:in `call'
vendor/bundle/gems/rack-1.6.13/lib/rack/head.rb:13:in `call'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/middleware/params_parser.rb:27:in `call'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/middleware/flash.rb:260:in `call'
vendor/bundle/gems/rack-1.6.13/lib/rack/session/abstract/id.rb:252:in `context'
vendor/bundle/gems/rack-1.6.13/lib/rack/session/abstract/id.rb:247:in `call'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/middleware/cookies.rb:560:in `call'
vendor/bundle/gems/activerecord-4.2.11.1/lib/active_record/query_cache.rb:36:in `call'
vendor/bundle/gems/activerecord-4.2.11.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:653:in `call'
vendor/bundle/gems/activerecord-4.2.11.1/lib/active_record/migration.rb:377:in `call'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
vendor/bundle/gems/activesupport-4.2.11.1/lib/active_support/callbacks.rb:88:in `__run_callbacks__'
vendor/bundle/gems/activesupport-4.2.11.1/lib/active_support/callbacks.rb:778:in `_run_call_callbacks'
vendor/bundle/gems/activesupport-4.2.11.1/lib/active_support/callbacks.rb:81:in `run_callbacks'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/middleware/callbacks.rb:27:in `call'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/middleware/reloader.rb:73:in `call'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/middleware/remote_ip.rb:78:in `call'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
vendor/bundle/gems/web-console-2.3.0/lib/web_console/middleware.rb:28:in `block in call'
vendor/bundle/gems/web-console-2.3.0/lib/web_console/middleware.rb:18:in `catch'
vendor/bundle/gems/web-console-2.3.0/lib/web_console/middleware.rb:18:in `call'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
vendor/bundle/gems/railties-4.2.11.1/lib/rails/rack/logger.rb:38:in `call_app'
vendor/bundle/gems/railties-4.2.11.1/lib/rails/rack/logger.rb:20:in `block in call'
vendor/bundle/gems/activesupport-4.2.11.1/lib/active_support/tagged_logging.rb:68:in `block in tagged'
vendor/bundle/gems/activesupport-4.2.11.1/lib/active_support/tagged_logging.rb:26:in `tagged'
vendor/bundle/gems/activesupport-4.2.11.1/lib/active_support/tagged_logging.rb:68:in `tagged'
vendor/bundle/gems/railties-4.2.11.1/lib/rails/rack/logger.rb:20:in `call'
vendor/bundle/gems/opentelemetry-instrumentation-rack-0.12.0/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb:67:in `block (3 levels) in call'
vendor/bundle/gems/opentelemetry-instrumentation-rack-0.12.0/lib/opentelemetry/instrumentation/rack.rb:45:in `block in with_span'
vendor/bundle/gems/opentelemetry-api-0.12.1/lib/opentelemetry/context.rb:62:in `with_value'
vendor/bundle/gems/opentelemetry-instrumentation-rack-0.12.0/lib/opentelemetry/instrumentation/rack.rb:45:in `with_span'
vendor/bundle/gems/opentelemetry-instrumentation-rack-0.12.0/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb:66:in `block (2 levels) in call'
vendor/bundle/gems/opentelemetry-api-0.12.1/lib/opentelemetry/trace/tracer.rb:29:in `block in in_span'
vendor/bundle/gems/opentelemetry-api-0.12.1/lib/opentelemetry/trace.rb:82:in `block in with_span'
vendor/bundle/gems/opentelemetry-api-0.12.1/lib/opentelemetry/context.rb:62:in `with_value'
vendor/bundle/gems/opentelemetry-api-0.12.1/lib/opentelemetry/trace.rb:82:in `with_span'
vendor/bundle/gems/opentelemetry-api-0.12.1/lib/opentelemetry/trace/tracer.rb:29:in `in_span'
vendor/bundle/gems/opentelemetry-instrumentation-rack-0.12.0/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb:63:in `block in call'
vendor/bundle/gems/opentelemetry-api-0.12.1/lib/opentelemetry/context.rb:46:in `with_current'
vendor/bundle/gems/opentelemetry-instrumentation-rack-0.12.0/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb:60:in `call'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/middleware/request_id.rb:21:in `call'
vendor/bundle/gems/rack-1.6.13/lib/rack/methodoverride.rb:22:in `call'
vendor/bundle/gems/rack-1.6.13/lib/rack/runtime.rb:18:in `call'
vendor/bundle/gems/activesupport-4.2.11.1/lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
vendor/bundle/gems/rack-1.6.13/lib/rack/lock.rb:17:in `call'
vendor/bundle/gems/actionpack-4.2.11.1/lib/action_dispatch/middleware/static.rb:120:in `call'
vendor/bundle/gems/rack-1.6.13/lib/rack/sendfile.rb:113:in `call'
vendor/bundle/gems/railties-4.2.11.1/lib/rails/engine.rb:518:in `call'
vendor/bundle/gems/railties-4.2.11.1/lib/rails/application.rb:165:in `call'
vendor/bundle/gems/rack-1.6.13/lib/rack/lock.rb:17:in `call'
vendor/bundle/gems/rack-1.6.13/lib/rack/content_length.rb:15:in `call'
vendor/bundle/gems/rack-1.6.13/lib/rack/handler/webrick.rb:88:in `service'
/Users/olone/.rbenv/versions/2.6.6/lib/ruby/2.6.0/webrick/httpserver.rb:140:in `service'
/Users/olone/.rbenv/versions/2.6.6/lib/ruby/2.6.0/webrick/httpserver.rb:96:in `run'
/Users/olone/.rbenv/versions/2.6.6/lib/ruby/2.6.0/webrick/server.rb:307:in `block in start_thread'
```
</details>

<details><summary>Gemfile.lock</summary>
```
GEM
  remote: https://rubygems.org/
  specs:
    actionmailer (4.2.11.1)
      actionpack (= 4.2.11.1)
      actionview (= 4.2.11.1)
      activejob (= 4.2.11.1)
      mail (~> 2.5, >= 2.5.4)
      rails-dom-testing (~> 1.0, >= 1.0.5)
    actionpack (4.2.11.1)
      actionview (= 4.2.11.1)
      activesupport (= 4.2.11.1)
      rack (~> 1.6)
      rack-test (~> 0.6.2)
      rails-dom-testing (~> 1.0, >= 1.0.5)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    actionview (4.2.11.1)
      activesupport (= 4.2.11.1)
      builder (~> 3.1)
      erubis (~> 2.7.0)
      rails-dom-testing (~> 1.0, >= 1.0.5)
      rails-html-sanitizer (~> 1.0, >= 1.0.3)
    activejob (4.2.11.1)
      activesupport (= 4.2.11.1)
      globalid (>= 0.3.0)
    activemodel (4.2.11.1)
      activesupport (= 4.2.11.1)
      builder (~> 3.1)
    activerecord (4.2.11.1)
      activemodel (= 4.2.11.1)
      activesupport (= 4.2.11.1)
      arel (~> 6.0)
    activesupport (4.2.11.1)
      i18n (~> 0.7)
      minitest (~> 5.1)
      thread_safe (~> 0.3, >= 0.3.4)
      tzinfo (~> 1.1)
    arel (6.0.4)
    binding_of_caller (1.0.0)
      debug_inspector (>= 0.0.1)
    builder (3.2.4)
    byebug (11.1.3)
    coffee-rails (4.1.1)
      coffee-script (>= 2.2.0)
      railties (>= 4.0.0, < 5.1.x)
    coffee-script (2.4.1)
      coffee-script-source
      execjs
    coffee-script-source (1.12.2)
    concurrent-ruby (1.1.8)
    connection_pool (2.2.3)
    crass (1.0.6)
    debug_inspector (1.0.0)
    erubis (2.7.0)
    execjs (2.7.0)
    ffi (1.14.2)
    globalid (0.4.2)
      activesupport (>= 4.2.0)
    i18n (0.9.5)
      concurrent-ruby (~> 1.0)
    jbuilder (2.9.1)
      activesupport (>= 4.2.0)
    jquery-rails (4.4.0)
      rails-dom-testing (>= 1, < 3)
      railties (>= 4.2.0)
      thor (>= 0.14, < 2.0)
    json (1.8.6)
    loofah (2.9.0)
      crass (~> 1.0.2)
      nokogiri (>= 1.5.9)
    mail (2.7.1)
      mini_mime (>= 0.1.1)
    mini_mime (1.0.2)
    mini_portile2 (2.5.0)
    minitest (5.14.3)
    nokogiri (1.11.1)
      mini_portile2 (~> 2.5.0)
      racc (~> 1.4)
    opentelemetry-api (0.12.1)
    opentelemetry-common (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-exporter-jaeger (0.12.0)
      opentelemetry-api (~> 0.12.0)
      opentelemetry-common (~> 0.12.0)
      thrift
    opentelemetry-instrumentation-active_model_serializers (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-all (0.12.1)
      opentelemetry-instrumentation-active_model_serializers (~> 0.12.0)
      opentelemetry-instrumentation-concurrent_ruby (~> 0.12.0)
      opentelemetry-instrumentation-dalli (~> 0.12.0)
      opentelemetry-instrumentation-delayed_job (~> 0.12.0)
      opentelemetry-instrumentation-ethon (~> 0.12.0)
      opentelemetry-instrumentation-excon (~> 0.12.0)
      opentelemetry-instrumentation-faraday (~> 0.12.0)
      opentelemetry-instrumentation-graphql (~> 0.12.0)
      opentelemetry-instrumentation-mongo (~> 0.12.0)
      opentelemetry-instrumentation-mysql2 (~> 0.12.0)
      opentelemetry-instrumentation-net_http (~> 0.12.0)
      opentelemetry-instrumentation-rack (~> 0.12.0)
      opentelemetry-instrumentation-rails (~> 0.12.0)
      opentelemetry-instrumentation-redis (~> 0.12.0)
      opentelemetry-instrumentation-restclient (~> 0.12.0)
      opentelemetry-instrumentation-ruby_kafka (~> 0.12.0)
      opentelemetry-instrumentation-sidekiq (~> 0.12.0)
      opentelemetry-instrumentation-sinatra (~> 0.12.0)
    opentelemetry-instrumentation-concurrent_ruby (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-dalli (0.12.0)
      opentelemetry-api (~> 0.12.0)
      opentelemetry-common (~> 0.12.0)
    opentelemetry-instrumentation-delayed_job (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-ethon (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-excon (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-faraday (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-graphql (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-mongo (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-mysql2 (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-net_http (0.12.0)
      opentelemetry-api (~> 0.12.0)
      opentelemetry-common (~> 0.12.0)
    opentelemetry-instrumentation-rack (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-rails (0.12.0)
      opentelemetry-api (~> 0.12.0)
      opentelemetry-instrumentation-rack (~> 0.12.0)
    opentelemetry-instrumentation-redis (0.12.0)
      opentelemetry-api (~> 0.12.0)
      opentelemetry-common (~> 0.12.0)
    opentelemetry-instrumentation-restclient (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-ruby_kafka (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-sidekiq (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-instrumentation-sinatra (0.12.0)
      opentelemetry-api (~> 0.12.0)
    opentelemetry-sdk (0.12.1)
      opentelemetry-api (~> 0.12.0)
      opentelemetry-common (~> 0.12.0)
    racc (1.5.2)
    rack (1.6.13)
    rack-protection (1.5.5)
      rack
    rack-test (0.6.3)
      rack (>= 1.0)
    rails (4.2.11.1)
      actionmailer (= 4.2.11.1)
      actionpack (= 4.2.11.1)
      actionview (= 4.2.11.1)
      activejob (= 4.2.11.1)
      activemodel (= 4.2.11.1)
      activerecord (= 4.2.11.1)
      activesupport (= 4.2.11.1)
      bundler (>= 1.3.0, < 2.0)
      railties (= 4.2.11.1)
      sprockets-rails
    rails-deprecated_sanitizer (1.0.3)
      activesupport (>= 4.2.0.alpha)
    rails-dom-testing (1.0.9)
      activesupport (>= 4.2.0, < 5.0)
      nokogiri (~> 1.6)
      rails-deprecated_sanitizer (>= 1.0.1)
    rails-html-sanitizer (1.3.0)
      loofah (~> 2.3)
    railties (4.2.11.1)
      actionpack (= 4.2.11.1)
      activesupport (= 4.2.11.1)
      rake (>= 0.8.7)
      thor (>= 0.18.1, < 2.0)
    rake (13.0.3)
    rb-fsevent (0.10.4)
    rb-inotify (0.10.1)
      ffi (~> 1.0)
    rdoc (4.3.0)
    redis (4.2.5)
    sass (3.7.4)
      sass-listen (~> 4.0.0)
    sass-listen (4.0.0)
      rb-fsevent (~> 0.9, >= 0.9.4)
      rb-inotify (~> 0.9, >= 0.9.7)
    sass-rails (5.0.7)
      railties (>= 4.0.0, < 6)
      sass (~> 3.1)
      sprockets (>= 2.8, < 4.0)
      sprockets-rails (>= 2.0, < 4.0)
      tilt (>= 1.1, < 3)
    sdoc (0.4.2)
      json (~> 1.7, >= 1.7.7)
      rdoc (~> 4.0)
    sidekiq (5.2.0)
      connection_pool (~> 2.2, >= 2.2.2)
      rack-protection (>= 1.5.0)
      redis (>= 3.3.5, < 5)
    sinatra (1.4.8)
      rack (~> 1.5)
      rack-protection (~> 1.4)
      tilt (>= 1.3, < 3)
    spring (2.1.1)
    sprockets (3.7.2)
      concurrent-ruby (~> 1.0)
      rack (> 1, < 3)
    sprockets-rails (3.2.2)
      actionpack (>= 4.0)
      activesupport (>= 4.0)
      sprockets (>= 3.0.0)
    sqlite3 (1.3.13)
    thor (1.1.0)
    thread_safe (0.3.6)
    thrift (0.13.0)
    tilt (2.0.10)
    turbolinks (5.2.1)
      turbolinks-source (~> 5.2)
    turbolinks-source (5.2.0)
    tzinfo (1.2.9)
      thread_safe (~> 0.1)
    uglifier (4.2.0)
      execjs (>= 0.3.0, < 3)
    web-console (2.3.0)
      activemodel (>= 4.0)
      binding_of_caller (>= 0.7.2)
      railties (>= 4.0)
      sprockets-rails (>= 2.0, < 4.0)

PLATFORMS
  ruby

DEPENDENCIES
  byebug
  coffee-rails (~> 4.1.0)
  jbuilder (~> 2.0)
  jquery-rails
  opentelemetry-api (~> 0.12.1)
  opentelemetry-exporter-jaeger (~> 0.12.0)
  opentelemetry-instrumentation-all (~> 0.12.1)
  opentelemetry-instrumentation-rack (~> 0.12.0)
  opentelemetry-instrumentation-rails (~> 0.12.0)
  opentelemetry-instrumentation-restclient (~> 0.12.0)
  opentelemetry-instrumentation-sidekiq (~> 0.12.0)
  opentelemetry-instrumentation-sinatra (~> 0.12.0)
  opentelemetry-sdk (~> 0.12.1)
  rails (= 4.2.11.1)
  sass-rails (~> 5.0)
  sdoc (~> 0.4.0)
  sidekiq (= 5.2)
  sinatra (= 1.4.8)
  spring
  sqlite3 (~> 1.3.6)
  turbolinks
  uglifier (>= 1.3.0)
  web-console (~> 2.0)

BUNDLED WITH
   1.17.3
```
</details>

**Comment by owais** [comment]
If updating the rails instrumentation to make the 3rd argument optional is an acceptable solution, I can submit a PR with that patch.

**Comment by fbogsany** [comment]
Rails 4.2.11 is very old at this point. The latest Rails release is 6.1.1, and only the 6.1.Z series is supported by the Rails core team. Our support extends back to 5.2.0:

https://github.com/open-telemetry/opentelemetry-ruby/blob/4a7054f36a734810f292971400ce4f07d49c0d26/instrumentation/rails/Appraisals#L8-L20

We will likely drop support for 5.2.0 soon, since it is broken with Ruby 3.0.

**Comment by johnnyshields** [comment]
@fbogsany Rails 5.x works on Ruby 2.6, which is slated for EOL on March 31, 2022. Support for Rails 5.2 should not be dropped before then.

I believe Rails 4.x works up to Ruby 2.5, which is EOL on March 31, 2021.

TLDR; it would be safe to add minimum version Rails >= 5, but this library should keep Rails 5 support around for sometime (especially given that it works today.)

**Comment by owais** [comment]
Makes sense @fbogsany. I don't think it's worth adding support for such an old version. Closing.
